### PR TITLE
Top Level Result Info

### PR DESCRIPTION
**Pull request by jessie-ross**
**This PR adds top level info to Experiment Results:**
- Top Level Stats
   - Displays max participant count over strategies
   - Display primary metric recommendation
- Health Traffic lights
   - Display sample ratio, sampler spammer ratio, crossovers
   - In debug mode show in depth details in added panel below
   - Link to new wiki page for details 

Future PRs (after the Calypso Client):
- Adding stopping time and Guard rail metrics

## How has this been tested?

- Automated tests that cover added/changed functionality
- Manual testing with production data (locally)

### CODE REVIEW

File: src/components/experiments/single-view/ExperimentDisableButton.tsx [hunk 1]
```diff
@@ -83,7 +83,7 @@ const ExperimentDisableButton = ({
         BackdropProps={{ className: clsx(isDisablingDangerous && dangerClasses.dangerBackdrop) }}
       >
         <DialogTitle>
-          <Typography variant='h5'>
+          <Typography variant='h5' component='h3'>
```
Comment (jessie-ross) [inline src/components/experiments/single-view/ExperimentDisableButton.tsx:86]: Fixing a warning

File: src/components/experiments/single-view/ExperimentRunButton.tsx [hunk 2]
```diff
@@ -86,7 +86,7 @@ const ExperimentRunButton = ({
         BackdropProps={{ className: dangerClasses.dangerBackdrop }}
       >
         <DialogTitle>
-          <Typography variant='h5'>
+          <Typography variant='h5' component='h3'>
```
Comment (jessie-ross) [inline src/components/experiments/single-view/ExperimentRunButton.tsx:89]: Fixing a warning

**Review by withinboredom — CHANGES_REQUESTED** [review]
All in all, this looks good. However, it may be worth taking a few lines at the beginning of the code path to assert sanity (remove `null` as a possible value) and would go a long way to cleaning up and making the code more readable/maintainable.

This is a big diff. It might be a good idea to do the traffic light calculations as a separate diff.

File: src/utils/math.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Assuming a distribution of X ~ Binomial(n, p), returns the probability of the number of trials being equal to x or more extreme than x.
+ *
+ * @param x number of successful trials
+ * @param n number of total trials
+ * @param p probability of success
+ */
+export function binomialProbValue(x: number, n: number, p: number): number {
+  const mean = n * p
+  const variance = n * p * (1 - p)
```
[the remark this answers is not in the document]
Comment (jessie-ross) [reply src/utils/math.ts:12]: I like the version in terms of `n` and `p` as it is more meaningful than the relationship between `variance` and `mean`.

File: src/utils/math.test.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { binomialProbValue } from './math'
+
+describe('utils/math.ts module', () => {
+  describe('binomialProbValue', () => {
+    it('is approximately correct', () => {
+      const aSmallValue = 0.000005
+      // Testing against known values
+      expect(Math.abs(binomialProbValue(1, 30, 0.5) - 0.000002)).toBeLessThan(aSmallValue)
```
[the remark this answers is not in the document]
Comment (jessie-ross) [reply src/utils/math.test.ts:8]: No I don't think so, it needs to be "small enough" for our intents and purposes but can't be too small as this is just an approximation.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx [hunk 5]
```diff
@@ -139,6 +206,100 @@ export default function ActualExperimentResults({
...
+  // ### Traffic light tests
+
+  // #### Sample Ratios
+
+  const variationHealthStats = experiment.variations.map((variation) => {
+    const p = variation.allocatedPercentage / 100
+
+    // User ratio
+    const nUsers = latestPrimaryMetricAnalysis?.participantStats['total']
+    const xUsers = latestPrimaryMetricAnalysis?.participantStats[`variation_${variation.variationId}`]
```
[the remark this answers is not in the document]
Comment (jessie-ross) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx:237]: Ah these are standard variables for a binomial trial with `x` successes out of `n` trials, I should probably document that/change them.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx [hunk 6]
```diff
@@ -139,6 +206,100 @@ export default function ActualExperimentResults({
...
+        crossoverRatio: xCrossovers / nCrossovers,
+      },
+    }
+  })
+
+  // istanbul ignore next; trivial
+  const userSampleRatioMinProbValue = _.min(variationHealthStats.map((x) => x.stats.usersRatioProbValue)) ?? 0
+  // istanbul ignore next; trivial
+  const userSampleRatioTrafficLight =
+    userSampleRatioMinProbValue < 0.01 ? TrafficLightColor.Red : TrafficLightColor.Green
```
[the remark this answers is not in the document]
Comment (jessie-ross) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx:281]: This particular case is a result of running `min` over a list - typescript can't check that a list has elements, and if there are no elements `min` returns `null`. The introduction of `?? 0` is purely to please typescript.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx [hunk 7]
```diff
@@ -139,6 +206,100 @@ export default function ActualExperimentResults({
     }),
   ]
 
+  // ### Top Level Stats
```
[the remark this answers is not in the document]
Comment (yanirs) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx:209]: Yeah, I think that a lot of these calculations should be pulled out to utility functions and unit tested.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (yanirs) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx:237]: Yeah, probably change it to more informative names. Names like `p`, `n`, and `x` have way too many meanings. 

Also, this introduces `users` as a new term, which we've been intentionally avoiding.

I'd go with:

* `p` -> `expectedVariationProbability`
* `nUsers` -> `totalParticipants`
* `xUsers` -> `variationParticipants`

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (yanirs) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx:281]: We can also please TypeScript by casting with `as <type>`, which is more self-documenting in this case.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx [hunk 8]
```diff
@@ -263,6 +479,20 @@ export default function ActualExperimentResults({
         }}
         detailPanel={DetailPanel}
       />
+      {
+        // istanbul ignore next; Debug only
```
[the remark this answers is not in the document]
Comment (yanirs) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx:483]: Debug mode should be tested, as it's also meant for experiment reviewers.

File: src/components/experiments/single-view/results/TrafficLightStatus.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      display: 'inline-block',
+      borderRadius: 100,
+      height: '.8em',
+      width: '.8em',
+      verticalAlign: 'middle',
+    },
+  }),
+)
+
+export default function TrafficLightStatus({ color, label }: { color: TrafficLightColor; label: string }): JSX.Element {
```
Comment (yanirs) [inline src/components/experiments/single-view/results/TrafficLightStatus.tsx:36]: I'm not colour blind and I find it hard to see the difference between the red and green here. We should have another visual indication and not rely on colour alone, e.g., :heavy_check_mark: :warning: :stop_sign: instead of colours.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx [hunk 10]
```diff
@@ -139,6 +206,100 @@ export default function ActualExperimentResults({
...
+      ] || 0) -
+      (primaryMetricLatestAnalysesByStrategy[AnalysisStrategy.MittNoSpammers]?.participantStats[
+        `variation_${variation.variationId}`
+      ] || 0)
+
+    // Crossovers
+    const nCrossovers =
+      (primaryMetricLatestAnalysesByStrategy[AnalysisStrategy.IttPure]?.participantStats['total'] || 0) -
+      (primaryMetricLatestAnalysesByStrategy[AnalysisStrategy.MittNoCrossovers]?.participantStats['total'] || 0)
+    const xCrossovers =
```
Comment (yanirs) [inline src/components/experiments/single-view/results/ActualExperimentResults.tsx:259]: As part of pulling this code out to helpers, it'd be good to also reduce duplication here as all these conditionals are very hard to parse.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx [hunk 11]
```diff
@@ -238,16 +393,77 @@ export default function ActualExperimentResults({
...
+                </div>
+              </>
+            )}
+          </Paper>
+          <Paper className={classes.summaryHealthPaper}>
+            <Link
+              href='https://github.com/Automattic/experimentation-platform/wiki/Experiment-Health#sample-user-ratios'
+              target='_blank'
+            >
+              <TrafficLightStatus color={userSampleRatioTrafficLight} label='Sample User Ratios' />
```
Comment (yanirs) [inline src/components/experiments/single-view/results/ActualExperimentResults.tsx:448]: As above, let's stick to talking about participants. Also, as this file is getting massive, it may be better to pull this part out to a separate component.

In addition, I think it'd be good to show more information on _why_ the label is what it is, e.g., change to `Participant SRM probability: <>`.

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (yanirs) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx:281]: +1, and please test the "trivial" cases, which would become easier when this is broken up to helpers and components.

**Review by yanirs — CHANGES_REQUESTED** [review]
I've only skimmed the code, but I agree with a lot of what Rob said.

Main top-level comments regarding the panel's content:

* We need a stronger visual indicator than colour
* It'd be good to include more information than the recommendation (i.e., the probabilities and ratios)
* We should also include an SRM warning for exposed participants

Main top-level comments regarding the code:

* This should be broken up and unit tested, which would make it easy to eliminate all/most of the `istanbul ignore` comments and also make the code more readable. Much of the logic can be separated from the component into helpers.
* As Rob noted, this is a pretty big PR, which would be easier to review if it's broken up.

File: src/components/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap [hunk 12]
```diff
@@ -8,14 +8,118 @@ exports[`renders correctly for 1 analysis datapoint 1`] = `
...
+            >
+              1,000
+            </h3>
+            <h6
+              class="MuiTypography-root MuiTypography-subtitle1"
+            >
+              <strong>
+                total participants
+              </strong>
+               as at
```
[the remark this answers is not in the document]
Comment (jessie-ross) [reply src/components/experiments/single-view/results/__snapshots__/ExperimentResults.test.tsx.snap:42]: Uhm, I think `as at` is better here as we want the connotation of a snapshot: https://english.stackexchange.com/a/112771

File: src/components/experiments/single-view/results/ActualExperimentResults.tsx [hunk 13]
```diff
@@ -139,6 +206,100 @@ export default function ActualExperimentResults({
...
+  const primaryMetricLatestAnalysesByStrategy = _.mapValues(
+    primaryMetricAssignmentAnalysesData.analysesByStrategyDateAsc,
+    _.last.bind(null),
+  )
+
+  const totalParticipants = Object.values(primaryMetricLatestAnalysesByStrategy)
+    .map(
+      (x) =>
+        // istanbul ignore next; trivial
+        x?.participantStats?.['total'] ?? 0,
```
[the remark this answers is not in the document]
Comment (jessie-ross) [reply src/components/experiments/single-view/results/ActualExperimentResults.tsx]: I've cleaned this up a bit (and removed the `participantsStats` check), see:
https://github.com/Automattic/abacus/pull/454/files#diff-6912e9b6b168daffa48dc35b58ada31ef27c57e60353cabbd01397c2a28f9966R212-R224

We are returning zero in either strategy missing or invalid data response (missing `total`). I don't think it is worth it to add two test cases just to get coverage here?

**Comment by jessie-ross** [comment]
## Update

- Removed the traffic light code.
- Refactored the Recommendation code:
   - Added `AggregateRecommendation` type.
   - Added `getAggregateRecommendation(analyses: Analyses[])`
   - Added `AggregateRecommendationDisplay` replacing `RecommendationString`.

There is a lot here but 95% of the changes now are refactoring the recommendation code, most of which are tests and the new functions and types which should be straight forward.

File: src/lib/analyses.ts [hunk 14]
```diff
@@ -19,3 +21,59 @@ export const RecommendationWarningToHuman = {
...
+export interface AggregateRecommendation {
+  type: AggregateRecommendationType
+  variationId?: number
+}
+
+/**
+ * Returns the aggregate recommendation over analyses of different analysis strategies.
+ * @param analyses Analyses of different strategies for the same day.
+ */
+export function getAggregateRecommendation(analyses: Analysis[]): AggregateRecommendation {
```
Comment (jessie-ross) [inline src/lib/analyses.ts:42]: The recommendation code is now all in once place and separate from the UI code.

**Comment by yanirs** [comment]
@jessie-ross it's been a long time since either @withinboredom or I looked at this PR. Given that we both suggested breaking it up back in January, it'd be good to separate refactoring changes to a different PR so this is easier to review and discuss.

**Comment by jessie-ross** [comment]
> Given that we both suggested breaking it up back in January, it'd be good to separate refactoring changes to a different PR so this is easier to review and discuss.

Yeah this is actually the broken up version, after moving AggregateRecommendation out to it's own file and refactoring for it, but sure I will break it up further. 

> Given the time that's passed and the changes made, I'm not sure what the PR is supposed to do at this point, but looking at Vercel, the big empty panel doesn't look right.

Yeah decided to leave it in as an intermediate change to the overall layout, but sure I can remove it.

**Comment by jessie-ross** [comment]
Closed by #471, with traffic lights to be included in a future PR
